### PR TITLE
Add multiple features, bug fixes and update libssh2 to the latest stable.

### PR DESCRIPTION
**Pull request by Red-M**
This is a new PR (from #90) to track the branch of changes for upstream as I see it now that I am required to make my own changes and publish under a new name as it appears the maintainer has disappeared.

The changes are as follows. I also recommend reading the commit messages and the changes on each commit instead of the roll up diff as it is HUGE.

### Changes

- Update libssh2 to 1.9.0 stable.
- Regenerate Cython files.
- Update .gitignore to include src directory to ignore libssh2 rebuilds.
- Add python 3.8 to appveyor.
- Add additional error codes.
- Add session.method_pref(), docs and tests.
- Add session.methods(), docs and tests.
- Add session.supported_algs(), docs and tests.
- Add session.flag(), docs and tests.
- Bug fix for session.userauth_publickey_frommemory() being provided no public key and throwing a type error.

### CODE REVIEW

**Comment by llchan** [comment]
Thanks for splitting off dedicated PR branch.  Could you revert the changes that rename `ssh2` to `redlibssh2` in this branch?  I assume the maintainer(s) will require that before it can be merged.

**Comment by Red-M** [comment]
Ah bugger, I did not see the 2nd line on that commit message, haha.
I'll see what I can do as I do not want to include someone else's source tree in the repository anymore, it just doesn't make sense to do that and its way better to only track fully released versions as well and this ensures that.
I'll take a look at splitting that commit.

**Comment by fruch** [comment]
seems like appveyor isn't happy
```
This repository is over its data quota. Account responsible for LFS bandwidth should purchase more data packs to restore access.
```
May I suggest trying to move to travis ? (they now support windows and mac)

**Comment by Red-M** [comment]
@llchan 
Sorry this took so long but I have been busy with other matters however this should be fine now.

@fruch 
Right now I am aiming for redlibssh2 to build packages for manylinux2014 on my own build infrastructure. Once I have packages I can publish to pypi I will be.
I'll also eventually get to travis builds because my build system easily supports that but for now, you should be able to install via my own repo if you'd like to use these patches and be sure you're getting them.

**Comment by fruch** [comment]
> @llchan
> Sorry this took so long but I have been busy with other matters however this should be fine now.
> 
> @fruch
> Right now I am aiming for redlibssh2 to build packages for manylinux2014 on my own build infrastructure. Once I have packages I can publish to pypi I will be.
> I'll also eventually get to travis builds because my build system easily supports that but for now, you should be able to install via my own repo if you'd like to use these patches and be sure you're getting them.

FYI you can build on travis for multiple platfrom quite easily with cibuildwheel
here's an example of a package i'm maintaining (it's now even has python3.9 wheels): 
https://travis-ci.org/github/scylladb/python-driver/builds/711061342
I'll be happy to give a hand with such a setup if you want

It's breaks my heart to see those kind of failures....
https://ci.appveyor.com/project/pkittenis/ssh2-python/builds/34304826/job/4yefexlvgd216bb7

**Comment by pkittenis** [comment]
Thanks for the PR. Have made similar changes, will aim to get these in.

If @Red-M  is interested can add you to this project. As has been obvious from this PR, there is only so much one maintainer can do.

**Comment by pkittenis** [comment]
| Thanks for splitting off dedicated PR branch. Could you revert the changes that rename ssh2 to redlibssh2 in this branch? I assume the maintainer(s) will require that before it can be merged.

Yes.

**Comment by pkittenis** [comment]
- [ ] Revert project renames

**Comment by pkittenis** [comment]
|  I'll be happy to give a hand with such a setup if you want [cibuildwheel]

@fruch , if you would like to take a stab at re-doing the wheel builds for travis CI would welcome a PR.

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "libssh2"]
```
Comment (pkittenis) [inline .gitmodules:1]: Would rather not use submodules. It adds quite a bit of overhead on the developer and can be confusing to work with. The source code is included in the repo so as not to have to use submodules.

**Comment by Red-M** [comment]
> Thanks for the PR. Have made similar changes, will aim to get these in.
> 
> If @Red-M is interested can add you to this project. As has been obvious from this PR, there is only so much one maintainer can do.

I'd love to do that.

> | Thanks for splitting off dedicated PR branch. Could you revert the changes that rename ssh2 to redlibssh2 in this branch? I assume the maintainer(s) will require that before it can be merged.
> 
> Yes.

This has already been done in the latest commit.

File: .gitmodules (hunk 1 above)
Comment (Red-M) [reply .gitmodules:1]: I'd prefer to use git submodules as they're designed to do this very thing for tracking a particular remote's branch.
In this PR I've told it to track the 1.9.0 release commit from libssh2's tags/release on github.

All this needs is some work on putting hit hooks into the repo with some minor bash scripts. (If libssh2 ever get around to releasing another version that is)

**Comment by pkittenis** [comment]
Closing this so can split the various tasks into different PRs. Libssh2 upgrade is being done in #101

In general better to do one thing per PR to make reviewing easier, eg API implementation, libssh2 upgrade, bug fixes etc. This PR has too many unrelated changes to effectively review at once.

**Comment by Red-M** [comment]
@pkittenis RE:
> If @Red-M is interested can add you to this project. As has been obvious from this PR, there is only so much one maintainer can do.
What would I need to do for this to be done?

As far as:
> can split the various tasks into different PRs
This is a little more difficult than you think as I removed the git commit history that contained LFS objects because it costs money from GitHub to use that and that carrying around packages and installing them is a bad practice instead of letting the OS for the builds handle that.
I'd need to redo the commits to build PRs from scratch.
I'd also need to delete my current fork and recreate it plus recreate my fork as its own repository as I can't create PRs to this repository from the current fork due the aforementioned LFS removal.

**Pull request closed without merging**